contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,144 @@
+    let string_lit = ::syn::parse2::<LitStr>(attr_stream2)
+        .map_err(|_| attr_span.error("expected string literal"))?;
+
+    let input = ::syn::parse::<DeriveInput>(input).unwrap();
+
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    if let Type::Path(type_path) = invocation.connection_type {
+        let inner_type = &type_path.path.segments.iter()
+            .map(|ps| ps.ident.clone())
+            .collect::<Vec<Ident>>();
+
+        let database_name = &invocation.db_name;
+        let handler_type = &invocation.type_name;
+        let pool_type = Ident::new(&format!("{}Pool", handler_type), Span::call_site());
+
+        let tokens = quote! {
+            use ::std::ops::*;
+            use ::rocket::http::*;
+            use ::rocket::request::*;
+            use ::rocket::*;
+            use ::rocket::fairing::*;
+            use ::rocket_contrib::*;
+            use ::rocket_contrib::databases::*;
+            pub struct #handler_type(r2d2::PooledConnection<<#(#inner_type)::* as Poolable>::Manager>);
+            pub struct #pool_type(r2d2::Pool<<#(#inner_type)::* as Poolable>::Manager>);
+
+            impl #handler_type {
+                pub fn pool(config: DatabaseConfig) -> Result<#pool_type, <#(#inner_type)::* as Poolable>::Error> {
+                    <#(#inner_type)::* as Poolable>::manager(&config)
+                        .and_then(|manager| {
+                             match r2d2::Pool::builder().max_size(config.pool_size).build(manager) {
+                                 Ok(pool) => Ok(pool),
+                                 Err(e) => Err(databases::DbError::PoolError(e)) // TODO: Fix this
+                             }
+                        })
+                        .and_then(|pool| Ok(#pool_type(pool)))
+                }
+
+                pub fn fairing() -> impl Fairing {
+                    AdHoc::on_attach(move |rocket| {
+                        if let Some(database_config) = databases::database_config(
+                            #database_name,
+                            &rocket.config().clone()
```

Comment:
Why are we cloning the config? That seems wasteful and unnecessary.

Reply:
You have to clone the config because otherwise you get an error when attempting to manage state via Rocket. The `config` method on `Rocket` takes itself by reference, so you can't move it within the same lexical scope (i.e. returning it from the `on_attach` method, if I understand the error correctly).

Looks like span information is incorrect, but you get the following error:
```
error[E0505]: cannot move out of `rocket` because it is borrowed
  --> contrib/lib/tests/databases.rs:14:5
   |
14 |     #[database("bar")]
   |     ^^^^^^^^^^^^^^^^^^
   |     |
   |     borrow of `rocket` occurs here
   |     move out of `rocket` occurs here
```